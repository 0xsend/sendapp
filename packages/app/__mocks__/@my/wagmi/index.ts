const mockMyWagmi = {
  baseMainnetClient: {
    chain: {
      id: 845337,
      name: 'Ethereum',
      nativeCurrency: {
        decimals: 18,
        name: 'Ethereum',
        symbol: 'ETH',
      },
    },
    transport: {
      url: 'http://localhost:8546',
    },
  },
  baseMainnet: {
    id: 845337,
    name: 'Base',
    nativeCurrency: {
      decimals: 18,
      name: 'Ethereum',
      symbol: 'ETH',
    },
    blockExplorers: {
      default: {
        name: 'Basescan',
        url: 'https://basescan.io',
      },
    },
  },
  usdcAddress: {
    845337: '0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913',
  },
  sendTokenAddress: {
    845337: '0xEab49138BA2Ea6dd776220fE26b7b8E446638956',
  },
  sendTokenV0Address: {
    845337: '0x3f14920c99BEB920Afa163031c4e47a3e03B3e4A',
  },
  spx6900Address: {
    845337: '0x50dA645f148798F68EF2d7dB7C1CB22A6819bb2C',
  },
  tokenPaymasterAddress: {
    845337: '0x5e421172B27658f2bD83BCBD13738ADdE00E7CA9',
  },
  entryPointAddress: {
    845337: '0x0000000071727De22E5E9d8BAf0edAc6f37da032',
  },
  sendtagCheckoutAddress: {
    845337: '0xfC1e51BBae1C1Ee9e6Cc629ea0023329EA5023a6',
  },
  sendVerifierAbi: [
    {
      type: 'function',
      inputs: [
        { name: 'message', internalType: 'bytes', type: 'bytes' },
        { name: 'signature', internalType: 'bytes', type: 'bytes' },
        { name: 'x', internalType: 'uint256', type: 'uint256' },
        { name: 'y', internalType: 'uint256', type: 'uint256' },
      ],
      name: 'verifySignature',
      outputs: [{ name: '', internalType: 'bool', type: 'bool' }],
      stateMutability: 'view',
    },
  ],
  sendVerifierProxyAddress: {
    845337: '0x6c38612d3f645711dd080711021fC1bA998a5628',
  },
  useWriteErc20Transfer: jest.fn().mockReturnValue({
    data: '0x123',
    writeContract: jest.fn(),
    isPending: false,
    error: null,
  }),
}

export const useWriteErc20Transfer = mockMyWagmi.useWriteErc20Transfer
export const baseMainnetClient = mockMyWagmi.baseMainnetClient
export const baseMainnet = mockMyWagmi.baseMainnet
export const usdcAddress = mockMyWagmi.usdcAddress
export const sendTokenAddress = mockMyWagmi.sendTokenAddress
export const sendTokenV0Address = mockMyWagmi.sendTokenV0Address
export const spx6900Address = mockMyWagmi.spx6900Address
export const tokenPaymasterAddress = mockMyWagmi.tokenPaymasterAddress
export const entryPointAddress = mockMyWagmi.entryPointAddress
export const sendVerifierAbi = mockMyWagmi.sendVerifierAbi
export const sendtagCheckoutAddress = mockMyWagmi.sendtagCheckoutAddress
export const sendVerifierProxyAddress = mockMyWagmi.sendVerifierProxyAddress
export default mockMyWagmi
