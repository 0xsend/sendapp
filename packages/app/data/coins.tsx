import { IconEthereum, IconSend, IconUSDC } from 'app/components/icons'
import {
  baseMainnet,
  usdcAddress as usdcAddresses,
  sendTokenAddress as sendAddresses,
} from '@my/wagmi'

export const coins = [
  { label: 'USDC', token: usdcAddresses[baseMainnet.id], icon: <IconUSDC size={'$2.5'} /> },
  { label: 'Ethereum', token: undefined, icon: <IconEthereum size={'$2.5'} /> },
  { label: 'Send', token: sendAddresses[baseMainnet.id], icon: <IconSend size={'$2.5'} /> },
]
