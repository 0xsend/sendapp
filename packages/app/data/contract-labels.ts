/**
 * Constants for contract labels used in the address book.
 * Using these constants ensures consistency across the codebase.
 */
export enum ContractLabels {
  SendEarn = 'Send Earn',
  Paymaster = 'Paymaster',
  SendtagCheckout = 'Sendtags',
  SendEarnAffiliate = 'Send Earn Affiliate',
}
