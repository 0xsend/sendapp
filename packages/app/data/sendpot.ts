export const SEND_POT_CONTRACT_ADDRESS = '0xa0A5611b9A1071a1D8A308882065c48650bAeE8b'

// Constant for basis points per ticket
export const BPS_PER_TICKET = 7000

/**
 * Calculates the actual number of tickets from the total basis points purchased.
 * @param bps - The total basis points representing tickets purchased.
 * @returns The actual number of tickets.
 */
export const calculateActualTickets = (bps: number): number => {
  if (bps <= 0 || BPS_PER_TICKET <= 0) {
    return 0
  }
  // Use Math.floor or Math.round depending on desired behavior for partial tickets
  return Math.floor(bps / BPS_PER_TICKET)
}
