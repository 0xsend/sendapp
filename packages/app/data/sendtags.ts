import { formatUnits, parseUnits } from 'viem'

export const maxNumSendTags = 5

/**
 * Returns the total price of all pending Sendtags.
 * @param pendingTags The pending Sendtags.
 * @returns The total price of all pending Sendtags.
 */
export function total(pendingTags: { name: string }[]) {
  return pendingTags.reduce((acc, { name }) => acc + price(name.length), BigInt(0))
}

/**
 * Returns the price of a Sendtag of the given length.
 * @param length The length of the Sendtag.
 * @returns The price of the Sendtag.
 */
export function price(length: number) {
  switch (length) {
    case 5:
      return parseUnits('15', 6) // 15 USDC
    case 4:
      return parseUnits('30', 6) // 30 USDC
    case 3:
    case 2:
    case 1:
      return parseUnits('60', 6) // 60 USDC
    default:
      return parseUnits('7', 6) // 7 USDC for 6+ characters
  }
}

/**
 * Returns the referral bonus for a Sendtag of the given length.
 * @param length The length of the Sendtag.
 * @returns The referral bonus for the Sendtag.
 */
export function bonus(length: number) {
  switch (length) {
    case 5:
      return parseUnits('4', 6) // 4 USDC
    case 4:
      return parseUnits('8', 6) // 8 USDC
    case 3:
    case 2:
    case 1:
      return parseUnits('16', 6) // 16 USDC
    default:
      return parseUnits('2', 6) // 2 USDC for 6+ characters
  }
}

/**
 * Returns the pricing information for Sendtags in USDC
 */
export const pricing = [
  {
    length: '6+ characters',
    price: formatUnits(price(6), 6),
  },
  {
    length: '5 characters',
    price: formatUnits(price(5), 6),
  },
  {
    length: '4 characters',
    price: formatUnits(price(4), 6),
  },
  {
    length: '1-3 characters',
    price: formatUnits(price(3), 6),
  },
]
