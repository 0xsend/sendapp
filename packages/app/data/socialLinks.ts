export const twitter = 'https://twitter.com/send'
export const telegram = 'https://t.me/send_app'
