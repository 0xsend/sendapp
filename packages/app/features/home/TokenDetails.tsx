import {
  AnimatePresence,
  Card,
  Paragraph,
  Separator,
  Spinner,
  Stack,
  Theme,
  XStack,
  YStack,
} from '@my/ui'
import type { CoinWithBalance } from 'app/data/coins'
import { ArrowDown, ArrowUp } from '@tamagui/lucide-icons'
import { IconError } from 'app/components/icons'
import { useTokenMarketData } from 'app/utils/coin-gecko'
import formatAmount from 'app/utils/formatAmount'
import { TokenActivity } from './TokenActivity'
import { useTokenPrices } from 'app/utils/useTokenPrices'
import { convertBalanceToFiat } from 'app/utils/convertBalanceToUSD'
import { IconCoin } from 'app/components/icons/IconCoin'

export function AnimateEnter({ children }: { children: React.ReactNode }) {
  return (
    <AnimatePresence>
      <Stack
        key="enter"
        animateOnly={['transform', 'opacity']}
        animation="200ms"
        enterStyle={{ opacity: 0, scale: 0.9 }}
        exitStyle={{ opacity: 0, scale: 0.95 }}
        opacity={1}
      >
        {children}
      </Stack>
    </AnimatePresence>
  )
}
export const TokenDetails = ({ coin }: { coin: CoinWithBalance }) => {
  return (
    <YStack f={1} gap="$5" $gtLg={{ w: '45%', pb: '$0' }} pb="$5">
      <YStack gap="$5">
        <Card p="$4.5" w={'100%'} jc={'space-between'} $gtLg={{ h: 244, p: '$6' }}>
          <YStack gap="$4">
            <XStack ai={'center'} gap={'$3'}>
              <IconCoin size={'$2'} symbol={coin.symbol} />
              <Paragraph
                size={'$7'}
                fontFamily={'$mono'}
                col={'$color12'}
                textTransform="uppercase"
                fontWeight={'700'}
              >
                {coin.label}
              </Paragraph>
            </XStack>
            <YStack gap={'$4'}>
              <TokenDetailsBalance coin={coin} />
              {coin.symbol !== 'USDC' && (
                <>
                  <Stack w={'100%'}>
                    <Separator bc={'$color10'} />
                  </Stack>
                  <TokenDetailsMarketData coin={coin} />
                </>
              )}
            </YStack>
          </YStack>
        </Card>
        {/* <XStack w={'100%'}>
          <LinkableButton href="/deposit" p="$7" mih={88} w="30%">
            <YStack gap="$2" jc={'space-between'} ai="center">
              <Theme name="green">
                <IconPlus
                  size={'$1.5'}
                  $theme-dark={{ color: '$color4' }}
                  $theme-light={{ color: '$color12' }}
                />
              </Theme>
              <Button.Text fontSize={'$4'} px="$2">
                Deposit
              </Button.Text>
            </YStack>
          </LinkableButton>
        </XStack> */}
      </YStack>
      <YStack gap={'$3'}>
        <TokenActivity coin={coin} />
      </YStack>
    </YStack>
  )
}

export const TokenDetailsMarketData = ({ coin }: { coin: CoinWithBalance }) => {
  const { data: tokenMarketData, isLoading: isLoadingMarketData } = useTokenMarketData(
    coin.coingeckoTokenId
  )

  const { data: prices, isLoading: isLoadingPrices } = useTokenPrices()

  const price = tokenMarketData?.at(0)?.current_price ?? prices?.[coin.token]

  const changePercent24h = tokenMarketData?.at(0)?.price_change_percentage_24h ?? null

  // Coingecko API returns a formatted price already. For now, we just want to make sure it doesn't have more than 8 digits
  // so the text doesn't get cut off.
  const formatPrice = (price: number) => price.toString().slice(0, 7)

  const formatPriceChange = (change: number) => {
    const fixedChange = change.toFixed(2)
    if (change > 0)
      return (
        <Theme name="green_active">
          <Paragraph fontSize="$4" fontWeight="500">{`${fixedChange}%`}</Paragraph>
          <ArrowUp size={'$0.9'} />
        </Theme>
      )
    if (change < 0)
      return (
        <Theme name="red_active">
          <Paragraph fontSize="$4" fontWeight="500">{`${fixedChange}%`}</Paragraph>
          <ArrowDown size={'$0.9'} />
        </Theme>
      )
    return <Paragraph fontSize="$4" fontWeight="500">{`${fixedChange}%`}</Paragraph>
  }

  if (isLoadingMarketData && isLoadingPrices) return <Spinner size="small" color={'$color12'} />

  return (
    <XStack gap="$3">
      {isLoadingPrices ? (
        <Spinner size="small" color={'$color12'} />
      ) : (
        <Paragraph
          fontSize={14}
          fontWeight="500"
          $theme-dark={{ color: '$gray8Light' }}
          color={'$color12'}
        >
          {price === undefined ? '' : `1 ${coin.symbol} = ${formatPrice(price)} USD`}
        </Paragraph>
      )}
      {isLoadingMarketData ? (
        <Spinner size="small" color={'$color12'} />
      ) : (
        <XStack gap={'$1.5'} ai="center" jc={'space-around'}>
          {changePercent24h === null ? (
            <XStack gap="$2" ai="center">
              <Paragraph color="$color10">Failed to load market data</Paragraph>
              <IconError size="$1.75" color={'$redVibrant'} />
            </XStack>
          ) : (
            formatPriceChange(changePercent24h)
          )}
        </XStack>
      )}
    </XStack>
  )
}

const TokenDetailsBalance = ({ coin }: { coin: CoinWithBalance }) => {
  const { data: tokenPrices, isLoading: isLoadingTokenPrices } = useTokenPrices()
  const { balance, decimals, formatDecimals = 5 } = coin

  if (coin.balance === undefined) {
    return <></>
  }

  const balanceInUSD = convertBalanceToFiat(coin, tokenPrices?.[coin.token])

  const balanceWithDecimals = Number(balance) / 10 ** (decimals ?? 0)
  const balanceWithDecimalsLength = balanceWithDecimals.toString().replace('.', '').length

  return (
    <XStack ai="flex-end" gap="$2">
      <Paragraph
        $platform-web={{ width: 'fit-content' }}
        $sm={{ fontSize: balanceWithDecimalsLength ? '$10' : 68, lineHeight: 38 }}
        fontSize={60}
        fontWeight={'900'}
        lineHeight={57}
        color={'$color12'}
      >
        {formatAmount(balanceWithDecimals.toString(), 10, formatDecimals)}
      </Paragraph>

      {coin.symbol !== 'USDC' ? (
        <Paragraph color={'$color10'} fontSize={'$3'} fontFamily={'$mono'}>
          {isLoadingTokenPrices || balanceInUSD === undefined
            ? ''
            : `($${formatAmount(balanceInUSD)})`}
        </Paragraph>
      ) : null}
    </XStack>
  )
}
