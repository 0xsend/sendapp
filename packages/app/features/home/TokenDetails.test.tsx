import '@jest/globals'

import { TamaguiProvider, config } from '@my/ui'
import { usdcCoin } from 'app/data/coins'
import { TokenDetails } from './TokenDetails'
import { act, render, screen } from '@testing-library/react-native'

jest.mock('app/features/home/utils/useTokenActivityFeed')

test('TokenDetails', async () => {
  jest.useFakeTimers()
  render(
    <TamaguiProvider defaultTheme={'dark'} config={config}>
      <TokenDetails coin={usdcCoin} />
    </TamaguiProvider>
  )
  await act(async () => {
    jest.advanceTimersByTime(2000)
    jest.runAllTimers()
  })
  expect(screen.toJSON()).toMatchSnapshot()

  expect(screen.getByText('Sent')).toBeOnTheScreen()
  expect(screen.getByText('Deposit')).toBeOnTheScreen()
  expect(screen.getByText('Received')).toBeOnTheScreen()
  expect(screen.getByText('/alice')).toBeOnTheScreen()
  expect(screen.getByText('0xa71CE00000000000000000000000000000000000')).toBeOnTheScreen()
  expect(screen.getByText('0x93F2FA7A16a7365e3895b0F6E6Ac7a832d6c761a')).toBeOnTheScreen()
  expect(screen.getByText('10 USDC')).toBeOnTheScreen()
  expect(screen.getByText('20 USDC')).toBeOnTheScreen()
  expect(screen.getByText('30 USDC')).toBeOnTheScreen()
})
