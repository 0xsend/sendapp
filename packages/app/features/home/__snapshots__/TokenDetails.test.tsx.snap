// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TokenDetails 1`] = `
<View
  style={
    {
      "flex": 1,
      "flexDirection": "column",
    }
  }
>
  <View
    style={
      {
        "flexDirection": "column",
        "marginTop": 18,
      }
    }
  >
    <Text
      cancelable={true}
      disabled={false}
      focusable={true}
      id=":r0:"
      minPressDuration={0}
      onBlur={[Function]}
      onFocus={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onPress={[Function]}
      onPressIn={[Function]}
      onPressOut={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "color": "#FFFFFF",
          "display": "flex",
          "fontFamily": "System",
          "fontSize": 19.2,
          "fontWeight": "500",
          "lineHeight": 44,
          "textTransform": "uppercase",
        }
      }
      suppressHighlighting={true}
      userSelect="none"
    >
      USDC BALANCE
    </Text>
    <Text
      accessibilityRole="header"
      selectable={true}
      style={
        {
          "color": "#FFFFFF",
          "fontFamily": "System",
          "fontSize": 68,
          "fontWeight": "700",
          "lineHeight": 57,
          "marginBottom": 0,
          "marginLeft": 0,
          "marginRight": 0,
          "marginTop": 0,
          "textTransform": "none",
        }
      }
      suppressHighlighting={true}
    >
      0
    </Text>
  </View>
  <View
    style={
      {
        "paddingBottom": 18,
        "paddingTop": 18,
        "width": "100%",
      }
    }
  >
    <View
      style={
        {
          "borderBottomColor": "#343434",
          "borderBottomWidth": 1,
          "borderLeftColor": "#343434",
          "borderLeftWidth": 0,
          "borderRightColor": "#343434",
          "borderRightWidth": 0,
          "borderStyle": "solid",
          "borderTopColor": "#343434",
          "borderTopWidth": 0,
          "flex": 1,
          "flexShrink": 0,
          "height": 0,
          "maxHeight": 0,
          "transform": [
            {
              "translateY": -0.5,
            },
          ],
        }
      }
    />
  </View>
  <View
    style={
      {
        "flexDirection": "column",
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "column",
          "gap": 24,
        }
      }
      testID="TokenDetailsHistory"
    >
      <Text
        accessibilityRole="header"
        selectable={true}
        style={
          {
            "color": "hsl(0, 0%, 42.5%)",
            "display": "none",
            "fontFamily": "System",
            "fontSize": 19.2,
            "fontWeight": "400",
            "lineHeight": 21,
            "marginBottom": 0,
            "marginLeft": 0,
            "marginRight": 0,
            "marginTop": 13,
          }
        }
        suppressHighlighting={true}
      >
        6/2/2024
      </Text>
      <View
        animatedStyle={
          {
            "value": {
              "opacity": 0,
              "transform": [
                {
                  "scale": 0.9,
                },
              ],
            },
          }
        }
        collapsable={false}
        forwardedRef={[Function]}
        style={
          {
            "opacity": 0,
            "transform": [
              {
                "scale": 0.9,
              },
            ],
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "borderBottomColor": "#3E4A3C",
              "borderBottomWidth": 1,
              "borderStyle": "solid",
              "flexDirection": "row",
              "gap": 18,
              "justifyContent": "space-between",
              "paddingBottom": 24,
              "width": "100%",
            }
          }
        >
          <View
            style={
              {
                "flex": 1,
                "flexDirection": "row",
                "gap": 21,
                "width": "100%",
              }
            }
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "borderBottomLeftRadius": 9,
                  "borderBottomRightRadius": 9,
                  "borderTopLeftRadius": 9,
                  "borderTopRightRadius": 9,
                  "flexDirection": "column",
                  "gap": 7,
                  "height": 48,
                  "justifyContent": "center",
                  "maxHeight": 48,
                  "maxWidth": 48,
                  "minHeight": 48,
                  "minWidth": 48,
                  "overflow": "hidden",
                  "position": "relative",
                  "width": 48,
                }
              }
            >
              <View
                style={
                  {
                    "bottom": 0,
                    "flexDirection": "column",
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                    "zIndex": 1,
                  }
                }
              >
                <Image
                  fullscreen={true}
                  onError={[Function]}
                  onLoad={[Function]}
                  source={
                    {
                      "uri": "https://ui-avatars.com/api/?name=0x93F2FA7A16a7365e3895b0F6E6Ac7a832d6c761a&size=256&format=png&background=86ad7f",
                    }
                  }
                  style={
                    {
                      "height": 48,
                      "width": 48,
                    }
                  }
                />
              </View>
              <View
                style={
                  {
                    "backgroundColor": "#86AE80",
                    "bottom": 0,
                    "flexDirection": "column",
                    "justifyContent": "center",
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                    "zIndex": 0,
                  }
                }
              >
                <View
                  style={
                    {
                      "alignItems": "center",
                      "borderBottomLeftRadius": 9,
                      "borderBottomRightRadius": 9,
                      "borderTopLeftRadius": 9,
                      "borderTopRightRadius": 9,
                      "flexDirection": "column",
                      "height": 48,
                      "justifyContent": "center",
                      "maxHeight": 48,
                      "maxWidth": 48,
                      "minHeight": 48,
                      "minWidth": 48,
                      "overflow": "hidden",
                      "position": "relative",
                      "width": 48,
                    }
                  }
                >
                  <View
                    style={
                      {
                        "bottom": 0,
                        "flexDirection": "column",
                        "left": 0,
                        "position": "absolute",
                        "right": 0,
                        "top": 0,
                        "zIndex": 1,
                      }
                    }
                  >
                    <Image
                      fullscreen={true}
                      onError={[Function]}
                      onLoad={[Function]}
                      source={
                        {
                          "uri": "https://ui-avatars.com/api/?name=0x93F2FA7A16a7365e3895b0F6E6Ac7a832d6c761a&size=256&format=png&background=86ad7f",
                        }
                      }
                      style={
                        {
                          "height": 48,
                          "width": 48,
                        }
                      }
                    />
                  </View>
                </View>
              </View>
            </View>
            <View
              style={
                {
                  "flex": 1,
                  "flexDirection": "column",
                  "gap": 4,
                  "overflow": "hidden",
                  "width": "100%",
                }
              }
            >
              <View
                style={
                  {
                    "flex": 1,
                    "flexDirection": "row",
                    "gap": 4,
                    "justifyContent": "space-between",
                    "width": "100%",
                  }
                }
              >
                <Text
                  style={
                    {
                      "color": "#FFFFFF",
                      "fontFamily": "System",
                      "fontSize": 24,
                    }
                  }
                  suppressHighlighting={true}
                >
                  Sent
                </Text>
                <Text
                  style={
                    {
                      "color": "#FFFFFF",
                      "fontFamily": "System",
                      "fontSize": 24,
                    }
                  }
                  suppressHighlighting={true}
                >
                  10 USDC
                </Text>
              </View>
              <View
                style={
                  {
                    "alignItems": "flex-start",
                    "flex": 1,
                    "flexDirection": "column",
                    "gap": 4,
                    "justifyContent": "space-between",
                    "overflow": "hidden",
                    "width": "100%",
                  }
                }
              >
                <Text
                  style={
                    {
                      "color": "#86AE80",
                      "fontFamily": "System",
                      "maxWidth": "100%",
                      "overflow": "hidden",
                    }
                  }
                  suppressHighlighting={true}
                >
                  0x93F2FA7A16a7365e3895b0F6E6Ac7a832d6c761a
                </Text>
                <Text
                  style={
                    {
                      "color": "#FFFFFF",
                      "fontFamily": "System",
                    }
                  }
                  suppressHighlighting={true}
                >
                  6/2/2024, 2:22:06 AM
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        animatedStyle={
          {
            "value": {
              "opacity": 0,
              "transform": [
                {
                  "scale": 0.9,
                },
              ],
            },
          }
        }
        collapsable={false}
        forwardedRef={[Function]}
        style={
          {
            "opacity": 0,
            "transform": [
              {
                "scale": 0.9,
              },
            ],
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "borderBottomColor": "#3E4A3C",
              "borderBottomWidth": 1,
              "borderStyle": "solid",
              "flexDirection": "row",
              "gap": 18,
              "justifyContent": "space-between",
              "paddingBottom": 24,
              "width": "100%",
            }
          }
        >
          <View
            style={
              {
                "flex": 1,
                "flexDirection": "row",
                "gap": 21,
                "width": "100%",
              }
            }
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "borderBottomLeftRadius": 9,
                  "borderBottomRightRadius": 9,
                  "borderTopLeftRadius": 9,
                  "borderTopRightRadius": 9,
                  "flexDirection": "column",
                  "gap": 7,
                  "height": 48,
                  "justifyContent": "center",
                  "maxHeight": 48,
                  "maxWidth": 48,
                  "minHeight": 48,
                  "minWidth": 48,
                  "overflow": "hidden",
                  "position": "relative",
                  "width": 48,
                }
              }
            >
              <View
                style={
                  {
                    "bottom": 0,
                    "flexDirection": "column",
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                    "zIndex": 1,
                  }
                }
              >
                <Image
                  fullscreen={true}
                  onError={[Function]}
                  onLoad={[Function]}
                  source={
                    {
                      "uri": "https://ui-avatars.com/api/?name=0xa71CE00000000000000000000000000000000000&size=256&format=png&background=86ad7f",
                    }
                  }
                  style={
                    {
                      "height": 48,
                      "width": 48,
                    }
                  }
                />
              </View>
              <View
                style={
                  {
                    "backgroundColor": "#86AE80",
                    "bottom": 0,
                    "flexDirection": "column",
                    "justifyContent": "center",
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                    "zIndex": 0,
                  }
                }
              >
                <View
                  style={
                    {
                      "alignItems": "center",
                      "borderBottomLeftRadius": 9,
                      "borderBottomRightRadius": 9,
                      "borderTopLeftRadius": 9,
                      "borderTopRightRadius": 9,
                      "flexDirection": "column",
                      "height": 48,
                      "justifyContent": "center",
                      "maxHeight": 48,
                      "maxWidth": 48,
                      "minHeight": 48,
                      "minWidth": 48,
                      "overflow": "hidden",
                      "position": "relative",
                      "width": 48,
                    }
                  }
                >
                  <View
                    style={
                      {
                        "bottom": 0,
                        "flexDirection": "column",
                        "left": 0,
                        "position": "absolute",
                        "right": 0,
                        "top": 0,
                        "zIndex": 1,
                      }
                    }
                  >
                    <Image
                      fullscreen={true}
                      onError={[Function]}
                      onLoad={[Function]}
                      source={
                        {
                          "uri": "https://ui-avatars.com/api/?name=0xa71CE00000000000000000000000000000000000&size=256&format=png&background=86ad7f",
                        }
                      }
                      style={
                        {
                          "height": 48,
                          "width": 48,
                        }
                      }
                    />
                  </View>
                </View>
              </View>
            </View>
            <View
              style={
                {
                  "flex": 1,
                  "flexDirection": "column",
                  "gap": 4,
                  "overflow": "hidden",
                  "width": "100%",
                }
              }
            >
              <View
                style={
                  {
                    "flex": 1,
                    "flexDirection": "row",
                    "gap": 4,
                    "justifyContent": "space-between",
                    "width": "100%",
                  }
                }
              >
                <Text
                  style={
                    {
                      "color": "#FFFFFF",
                      "fontFamily": "System",
                      "fontSize": 24,
                    }
                  }
                  suppressHighlighting={true}
                >
                  Received
                </Text>
                <Text
                  style={
                    {
                      "color": "#FFFFFF",
                      "fontFamily": "System",
                      "fontSize": 24,
                    }
                  }
                  suppressHighlighting={true}
                >
                  20 USDC
                </Text>
              </View>
              <View
                style={
                  {
                    "alignItems": "flex-start",
                    "flex": 1,
                    "flexDirection": "column",
                    "gap": 4,
                    "justifyContent": "space-between",
                    "overflow": "hidden",
                    "width": "100%",
                  }
                }
              >
                <Text
                  style={
                    {
                      "color": "#86AE80",
                      "fontFamily": "System",
                      "maxWidth": "100%",
                      "overflow": "hidden",
                    }
                  }
                  suppressHighlighting={true}
                >
                  0xa71CE00000000000000000000000000000000000
                </Text>
                <Text
                  style={
                    {
                      "color": "#FFFFFF",
                      "fontFamily": "System",
                    }
                  }
                  suppressHighlighting={true}
                >
                  6/2/2024, 2:22:02 AM
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        animatedStyle={
          {
            "value": {
              "opacity": 0,
              "transform": [
                {
                  "scale": 0.9,
                },
              ],
            },
          }
        }
        collapsable={false}
        forwardedRef={[Function]}
        style={
          {
            "opacity": 0,
            "transform": [
              {
                "scale": 0.9,
              },
            ],
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "borderBottomColor": "#3E4A3C",
              "borderBottomWidth": 1,
              "borderStyle": "solid",
              "flexDirection": "row",
              "gap": 18,
              "justifyContent": "space-between",
              "paddingBottom": 24,
              "width": "100%",
            }
          }
        >
          <View
            style={
              {
                "flex": 1,
                "flexDirection": "row",
                "gap": 21,
                "width": "100%",
              }
            }
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "borderBottomLeftRadius": 9,
                  "borderBottomRightRadius": 9,
                  "borderTopLeftRadius": 9,
                  "borderTopRightRadius": 9,
                  "flexDirection": "column",
                  "gap": 7,
                  "height": 48,
                  "justifyContent": "center",
                  "maxHeight": 48,
                  "maxWidth": 48,
                  "minHeight": 48,
                  "minWidth": 48,
                  "overflow": "hidden",
                  "position": "relative",
                  "width": 48,
                }
              }
            >
              <View
                style={
                  {
                    "bottom": 0,
                    "flexDirection": "column",
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                    "zIndex": 1,
                  }
                }
              >
                <Image
                  fullscreen={true}
                  onError={[Function]}
                  onLoad={[Function]}
                  source={
                    {
                      "uri": "https://i.pravatar.cc/500?u=alice",
                    }
                  }
                  style={
                    {
                      "height": 48,
                      "width": 48,
                    }
                  }
                />
              </View>
              <View
                style={
                  {
                    "backgroundColor": "#86AE80",
                    "bottom": 0,
                    "flexDirection": "column",
                    "justifyContent": "center",
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                    "zIndex": 0,
                  }
                }
              >
                <View
                  style={
                    {
                      "alignItems": "center",
                      "borderBottomLeftRadius": 9,
                      "borderBottomRightRadius": 9,
                      "borderTopLeftRadius": 9,
                      "borderTopRightRadius": 9,
                      "flexDirection": "column",
                      "height": 48,
                      "justifyContent": "center",
                      "maxHeight": 48,
                      "maxWidth": 48,
                      "minHeight": 48,
                      "minWidth": 48,
                      "overflow": "hidden",
                      "position": "relative",
                      "width": 48,
                    }
                  }
                >
                  <View
                    style={
                      {
                        "bottom": 0,
                        "flexDirection": "column",
                        "left": 0,
                        "position": "absolute",
                        "right": 0,
                        "top": 0,
                        "zIndex": 1,
                      }
                    }
                  >
                    <Image
                      fullscreen={true}
                      onError={[Function]}
                      onLoad={[Function]}
                      source={
                        {
                          "uri": "https://ui-avatars.com/api/?name=Alice&size=256&format=png&background=86ad7f",
                        }
                      }
                      style={
                        {
                          "height": 48,
                          "width": 48,
                        }
                      }
                    />
                  </View>
                </View>
              </View>
            </View>
            <View
              style={
                {
                  "flex": 1,
                  "flexDirection": "column",
                  "gap": 4,
                  "overflow": "hidden",
                  "width": "100%",
                }
              }
            >
              <View
                style={
                  {
                    "flex": 1,
                    "flexDirection": "row",
                    "gap": 4,
                    "justifyContent": "space-between",
                    "width": "100%",
                  }
                }
              >
                <Text
                  style={
                    {
                      "color": "#FFFFFF",
                      "fontFamily": "System",
                      "fontSize": 24,
                    }
                  }
                  suppressHighlighting={true}
                >
                  Received
                </Text>
                <Text
                  style={
                    {
                      "color": "#FFFFFF",
                      "fontFamily": "System",
                      "fontSize": 24,
                    }
                  }
                  suppressHighlighting={true}
                >
                  30 USDC
                </Text>
              </View>
              <View
                style={
                  {
                    "alignItems": "flex-start",
                    "flex": 1,
                    "flexDirection": "column",
                    "gap": 4,
                    "justifyContent": "space-between",
                    "overflow": "hidden",
                    "width": "100%",
                  }
                }
              >
                <Text
                  style={
                    {
                      "color": "#86AE80",
                      "fontFamily": "System",
                      "maxWidth": "100%",
                      "overflow": "hidden",
                    }
                  }
                  suppressHighlighting={true}
                >
                  /alice
                </Text>
                <Text
                  style={
                    {
                      "color": "#FFFFFF",
                      "fontFamily": "System",
                    }
                  }
                  suppressHighlighting={true}
                >
                  6/2/2024, 2:21:44 AM
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        animatedStyle={
          {
            "value": {
              "opacity": 0,
              "transform": [
                {
                  "scale": 0.9,
                },
              ],
            },
          }
        }
        collapsable={false}
        forwardedRef={[Function]}
        style={
          {
            "opacity": 0,
            "transform": [
              {
                "scale": 0.9,
              },
            ],
          }
        }
      />
    </View>
  </View>
</View>
`;
